### ISSUE TITLE
[HELP] Windows install: "cannot open input file 'cairo-gobject.lib'"

### ISSUE THREAD

**Issue by pieterdd** [gtk-rs/gtk4-rs#1257]
When following the [Windows installation instructions](https://gtk-rs.org/gtk4-rs/stable/latest/book/installation_windows.html), I ran into a build problem during the librsvg steps at this command:

```
nmake /f Makefile.vc CFG=release install PREFIX=C:\gnome
```

Sanity checks I did based on a checklist by @wroyca:

- `git --version` outputs 'git version 2.37.2.windows.2'
- I found `C:\pkg-config-lite-0.28-1\bin` in the output of `echo $Env:Path`, same with `C:\gnome\bin`
- `echo $Env:PKG_CONFIG_PATH` outputs `C:\gnome\lib\pkgconfig`
- I'm using the PowerShell version of Visual Studio 2022 Developer Command Prompt v17.4.4. When I run the command in the legacy command prompt, I get the same error.

The error is:

```
  = note: LINK : fatal error LNK1181: cannot open input file 'cairo-gobject.lib'


The following warnings were emitted during compilation:

warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-pdf, but could not extract library base name to pass to linker command line
warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-ps, but could not extract library base name to pass to linker command line

error: could not compile `librsvg` due to previous error

Caused by:
  process didn't exit successfully: `rustc --crate-name rsvg_convert (...)
```

I could not find `cairo-gobject.lib` in either C:\gnome or C:\gtk. I'm not sure in which folder I should expect it. Do you think the presence of MSYS64 on my system is somehow interfering here? `C:/msys64/ucrt64/lib/` is not on my path but `C:\msys64\ucrt64\bin` is.

**Comment by wroyca** [comment]
Yes, it seems that the presence of MSYS64 is interfering, best to remove it temporarily from path and see if you can continue without it. It is also worth checking that you have gettext 0.21 and iconv 1.16 (static) properly installed and available in your path as well

`GETTEXTCLDRDIR C:\Program Files\gettext-iconv\lib\gettext`

**Comment by pieterdd** [comment]
Having removed the msys64 bin folder from my path, installed the correct static 64-bit variant of gettext 0.21, restarted the terminal app, verified that `echo $Env:Path` no longer contains it, checked that `echo $Env:GETTEXTCLDRDIR` equals `C:\Program Files\gettext-iconv\lib\gettext`, and checked that `xgettext --version` mentions 0.21, I still got the error including the msys64 references:

```
warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-pdf, but could not extract library base name to pass to linker command line
warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-ps, but could not extract library base name to pass to linker command line
```

Then I recreated the librsvg folder from scratch, ran the steps related to it again, and up came the "fatal error LNK1181: cannot open input file 'cairo-gobject.lib'" error once more. I feel like there's still a 50% chance of it being something silly I forgot to do or check, I'm just not sure where to look for it.

**Comment by wroyca** [comment]
> Having removed the msys64 bin folder from my path, installed the correct static 64-bit variant of gettext 0.21, restarted the terminal app, verified that `echo $Env:Path` no longer contains it, checked that `echo $Env:GETTEXTCLDRDIR` equals `C:\Program Files\gettext-iconv\lib\gettext`, and checked that `xgettext --version` mentions 0.21, I still got the error including the msys64 references:
> 
> ```
> warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-pdf, but could not extract library base name to pass to linker command line
> warning: File path C:/msys64/ucrt64/lib/libz.dll.a found in pkg-config file for cairo-ps, but could not extract library base name to pass to linker command line
> ```
> 
> Then I recreated the librsvg folder from scratch, ran the steps related to it again, and up came the "fatal error LNK1181: cannot open input file 'cairo-gobject.lib'" error once more. I feel like there's still a 50% chance of it being something silly I forgot to do or check, I'm just not sure where to look for it.

Hm, that's strange, can you run `where pkg-config` and can you pass the log from `meson setup builddir --prefix=C:/gnome -Dbuild-tests=false -Dmedia-gstreamer=disabled` in C:/gtk - this will give us a better insight into what might possibly be going wrong ( be sure to wipe builddir/ beforehand. )

It is possible that mingw pkgconfig takes precedence over pkgconfig-lite, which causes the problems mentioned

**Comment by pieterdd** [comment]
**tl;dr** Thanks, I got it up and running with your help! I did jump through a few hoops along the way. I'll pay it forward by leaving my process here so Google can crawl it for future use.

---

Interesting: a pkg-config binary from another folder (not one related to MSYS64/MinGW though) indeed took precedence. So I moved up the folder in the Path variable and ran this sanity check:

```
PS C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools> (Get-Command pkg-config).Path
C:\pkg-config-lite-0.28-1\bin\pkg-config.exe
```

I wiped the gnome folder, recreated the gtk/libxml2/librsvg directories using git clone, then started again. Curiously, the build was failing earlier than before.

```
PS C:\gtk> meson setup builddir --prefix=C:/gnome -Dbuild-tests=false -Dmedia-gstreamer=disabled
The Meson build system
Version: 1.0.0
Source dir: C:\gtk
Build dir: C:\gtk\builddir
Build type: native build
Project name: gtk
Project version: 4.9.3
C compiler for the host machine: cl (msvc 19.34.31937 "Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64")
C linker for the host machine: link link 14.34.31937.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
cl.exe does not actually support gnu standards, and meson will instead demote to the nearest ISO C standard. This may cause compilation to fail.
Has header "crt/externs.h" : NO
Has header "dev/evdev/input.h" : NO
Has header "dlfcn.h" : NO
Has header "ftw.h" : NO
Has header "inttypes.h" : YES
Has header "linux/input.h" : NO
Has header "linux/memfd.h" : NO
Has header "locale.h" : YES
Has header "memory.h" : YES
Has header "stdint.h" : YES
Has header "stdlib.h" : YES
Has header "strings.h" : NO
Has header "string.h" : YES
Has header "sys/mman.h" : NO
Has header "sys/param.h" : NO
Has header "sys/stat.h" : YES
Has header "sys/sysinfo.h" : NO
Has header "sys/systeminfo.h" : NO
Has header "sys/time.h" : NO
Has header "sys/types.h" : YES
Has header "unistd.h" : NO
Has header "gio/gdesktopappinfo.h" : NO
Library m found: NO
Checking for function "dcgettext" with dependency -lm: NO
Checking for function "getpagesize" with dependency -lm: NO
Checking for function "getresuid" with dependency -lm: NO
Checking for function "lstat" with dependency -lm: NO
Checking for function "memfd_create" with dependency -lm: NO
Checking for function "mmap" with dependency -lm: NO
Checking for function "posix_fallocate" with dependency -lm: NO
Checking for function "_lock_file" with dependency -lm: YES
Checking for function "flockfile" with dependency -lm: NO
Checking for function "mkstemp" with dependency -lm: NO
Checking for function "mallinfo2" with dependency -lm: NO
Checking for function "sincos" with dependency -lm: NO
Checking for function "sincosf" with dependency -lm: NO
Checking if "sigsetjmp" : links: NO
Checking if "__uint128_t available" compiles: NO
Checking for function "mlock" : NO
Checking for function "madvise" : NO
Compiler for C supports arguments -FImsvc_recommended_pragmas.h: NO
Compiler for C supports arguments -D_USE_MATH_DEFINES: YES
Compiler for C supports arguments -utf-8: YES
Compiler for C supports arguments -Zc:preprocessor: YES
meson.build:263: WARNING: add_languages is missing native:, assuming languages are wanted for both host and build.
C++ compiler for the host machine: cl (msvc 19.34.31937 "Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64")
C++ linker for the host machine: link link 14.34.31937.0
Compiler for C++ supports arguments -FImsvc_recommended_pragmas.h: NO
Compiler for C++ supports arguments -D_USE_MATH_DEFINES: YES
Compiler for C++ supports arguments -utf-8: YES
Compiler for C++ supports arguments -Zc:preprocessor: YES
Compiler for C supports arguments -FImsvc_recommended_pragmas.h: NO (cached)
Compiler for C supports arguments -D_USE_MATH_DEFINES: YES (cached)
Compiler for C supports arguments -utf-8: YES (cached)
Compiler for C supports arguments -Zc:preprocessor: YES (cached)
Compiler for C supports link arguments -Wl,-Bsymbolic: NO
Compiler for C supports link arguments -Wl,-z,relro: NO
Compiler for C supports link arguments -Wl,-z,now: NO
Found pkg-config: C:\pkg-config-lite-0.28-1\bin\pkg-config.EXE (0.28)
Run-time dependency glib-2.0 found: YES 2.74.4
Run-time dependency gobject-2.0 found: YES 2.74.4
Run-time dependency gio-windows-2.0 found: YES 2.74.4
Run-time dependency gmodule-2.0 found: YES 2.74.4
Run-time dependency cairo found: YES 1.17.6
Run-time dependency cairo-gobject found: YES 1.17.6
Run-time dependency pango found: YES 1.50.12
Run-time dependency fribidi found: YES 1.0.12
Run-time dependency harfbuzz found: YES 4.0.0
Run-time dependency pangoft2 found: YES 1.50.12
Run-time dependency pangowin32 found: YES 1.50.12
Run-time dependency pangocairo found: YES 1.50.12
Run-time dependency gdk-pixbuf-2.0 found: YES 2.42.11
Found CMake: C:\Program Files\CMake\bin\cmake.EXE (3.25.1)
Run-time dependency libpng found: NO (tried pkgconfig and cmake)
Run-time dependency png found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency libpng
Downloading libpng source from https://github.com/glennrp/libpng/archive/v1.6.37.tar.gz
Downloading file of unknown size.
Downloading libpng patch from https://wrapdb.mesonbuild.com/v2/libpng_1.6.37-3/get_patch
Downloading: 100%|██████████████████████████████████████| 2024/2024 [00:00<00:00, 2029469.59bytes/s]

Executing subproject libpng

libpng| Project name: libpng
libpng| Project version: 1.6.37
libpng| C compiler for the host machine: cl (msvc 19.34.31937 "Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64")
libpng| C linker for the host machine: link link 14.34.31937.0
libpng| Fetching value of define "__ANDROID__" :
libpng| Run-time dependency zlib found: NO (tried pkgconfig, cmake and system)

subprojects\libpng-1.6.37\meson.build:17:2: ERROR: Dependency "zlib" not found, tried pkgconfig, cmake and system
```

Did another wipe and then the error stopped occurring, which I found odd.  Not sure if a dependency silently failed to download or whether I subconsciously executed the procedure incorrectly. I got past the `cairo-gobject.lib` error and got this in the `nmake /f Makefile.vc CFG=release install PREFIX=C:\gnome` step:

```
=========================
Configuration for librsvg
=========================
Build Type: release

Built Tools:
rsvg-convert

Introspection: no
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) moved.

[main 2023-01-15T09:33:08.733Z] update#setState idle
[main 2023-01-15T09:33:09.911Z] Starting extension host with pid 3080 (fork() took 12 ms).
[main 2023-01-15T09:33:17.337Z] Waiting for extension host with pid 3080 to exit.
[main 2023-01-15T09:33:17.374Z] Extension host with pid 3080 exited with code: 0, signal: null.
*** librsvg-2.0.pc is not generated! Generate it later using $(PYTHON) --version=2.55.90 --prefix=$(PREFIX)
```

At this point a VSCode window for C:\librsvg\win32\rsvgpc.py opened, when I assume instead it intended to execute the script. I tried both `nmake /f generate-msvc.mak generate-nmake-files PYTHON="C:\Python310\python.exe"` and `nmake /f generate-msvc.mak generate-nmake-files PYTHON=C:\Python310\python.exe` with the same outcome. The next time, I tried to inspect the spawned process:

```
[main 2023-01-15T09:46:16.106Z] update#setState idle
[main 2023-01-15T09:46:17.002Z] Starting extension host with pid 14848 (fork() took 14 ms).
[main 2023-01-15T09:46:46.118Z] update#setState checking for updates
[main 2023-01-15T09:46:46.143Z] update#setState idle
```

According to Process Explorer, the command line path for PID 14847 was `"C:\Users\Username\AppData\Local\Programs\Microsoft VS Code\Code.exe" --ms-enable-electron-run-as-node --inspect-port=0 "c:\Users\Username\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork" --type=extensionHost --skipWorkspaceStorageLock`. Its parent process was Code.exe (PID 12176), whose command line invocation was `"C:\Users\Username\AppData\Local\Programs\Microsoft VS Code\Code.exe" "C:\librsvg\win32\rsvgpc.py"`.

As a workaround, I took a gamble and executed `python "C:\librsvg\win32\rsvgpc.py" --version=2.55.90 --prefix=C:\gnome` in a different VS 2022 PowerShell. Then I quit the VSCode process and got this:

```
[main 2023-01-15T09:46:16.106Z] update#setState idle
[main 2023-01-15T09:46:17.002Z] Starting extension host with pid 14848 (fork() took 14 ms).
[main 2023-01-15T09:46:46.118Z] update#setState checking for updates
[main 2023-01-15T09:46:46.143Z] update#setState idle
[3664:0115/104816.045:ERROR:gpu_init.cc(481)] Passthrough is not supported, GL is disabled, ANGLE is
[main 2023-01-15T09:54:16.809Z] Waiting for extension host with pid 14848 to exit.
[main 2023-01-15T09:54:16.831Z] Extension host with pid 14848 exited with code: 0, signal: null.
        1 file(s) moved.
```

I compiled a test application and now have a Gtk4 window on my desktop! Thanks for your help.

**Comment by wroyca** [comment]
Glad I could help! :) 

- Unfortunately, yes, pkgconfig has always been hit-or-miss on Windows, too many unrelated things can inadvertently affect it.

@bilelmoussaoui Should we add a notice to tell users to make sure that nothing override pkgconfig-lite? This can be quite sneaky as sometimes a random install will pack its own pkgconfig.

**Comment by bilelmoussaoui** [comment]
@wroyca sure, anything that would help users debug an installation issue would be very helpful
